### PR TITLE
[SPARK-22671][SQL] Make SortMergeJoin shuffle read less data when wholeStageCodegen is off

### PR DESCRIPTION
**Pull request by gcz2022**
## What changes were proposed in this pull request?

In SortMergeJoin(with wholeStageCodegen), an optimization already exists: if the left table of a partition is empty then there is no need to read the right table of this corresponding partition. This benefits the case in which many partitions of left table is empty and the right table is big.

While in the code path without wholeStageCodegen, this optimization doesn't happen. This is mainly due to the lack of optimization in codegen-SortMergeJoin & the lack of support in `SortExec`, which doesn't do lazy evaluation. UI of TPC-DS q37 when wholeStageCodegen is off:
off_wholestage_before

When the switch is on: 
wholestage_on

This PR lazy evaluates sort, and only trigger the right table read after reading the left table and find it's not empty.

After this PR, with wholeStageCodegen off:
wholestage_off

## How was this patch tested?

Existed test suites.

### CODE REVIEW

**Comment by gczsjdy** [comment]
cc @cloud-fan @viirya @ConeyLiu

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 1]
```diff
@@ -699,39 +700,44 @@ private[joins] class SortMergeJoinScanner(
...
-          assert(!bufferedRowKey.anyNull)
-          comp = keyOrdering.compare(streamedRowKey, bufferedRowKey)
-          if (comp > 0) advancedBufferedToRowWithNullFreeJoinKey()
-          else if (comp < 0) advancedStreamed()
-        }
-      } while (streamedRow != null && bufferedRow != null && comp != 0)
-      if (streamedRow == null || bufferedRow == null) {
-        // We have either hit the end of one of the iterators, so there can be no more matches.
+      // To make sure vars like bufferedRow is set
+      advancedBufferedIterRes
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: To advance buffer iterator here, won't we miss the `bufferedRow` advanced before?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 2]
```diff
@@ -750,6 +756,8 @@ private[joins] class SortMergeJoinScanner(
       bufferedMatches.clear()
       false
     } else {
+      // To make sure vars like bufferedRow is set
+      advancedBufferedIterRes
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: ditto. We may miss the bufferedRow advanced before.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 3]
```diff
@@ -674,8 +674,9 @@ private[joins] class SortMergeJoinScanner(
   private[this] val bufferedMatches =
     new ExternalAppendOnlyUnsafeRowArray(inMemoryThreshold, spillThreshold)
 
-  // Initialization (note: do _not_ want to advance streamed here).
-  advancedBufferedToRowWithNullFreeJoinKey()
+  // Initialization (note: do _not_ want to advance streamed here). This is made lazy to prevent
+  // unnecessary trigger of calculation.
+  private lazy val advancedBufferedIterRes = advancedBufferedToRowWithNullFreeJoinKey()
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Isn't it the same to simply call `advancedBufferedToRowWithNullFreeJoinKey` at needed places?

Comment (gczsjdy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: This function should be called (to try to set `BufferedRow`) before `BufferedRow` is checked, and it should be only once. This is the original requirement due to the logic. While to add this optimization, I think this is the best way.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 1 above)
Comment (gczsjdy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: This advance function is only called once actually, so no `bufferedRow` will be missed. Or maybe I didn't understand your meaning?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 4]
```diff
@@ -699,39 +700,44 @@ private[joins] class SortMergeJoinScanner(
...
-          comp = keyOrdering.compare(streamedRowKey, bufferedRowKey)
-          if (comp > 0) advancedBufferedToRowWithNullFreeJoinKey()
-          else if (comp < 0) advancedStreamed()
-        }
-      } while (streamedRow != null && bufferedRow != null && comp != 0)
-      if (streamedRow == null || bufferedRow == null) {
-        // We have either hit the end of one of the iterators, so there can be no more matches.
+      // To make sure vars like bufferedRow is set
+      advancedBufferedIterRes
+      if (matchJoinKey != null && keyOrdering.compare(streamedRowKey, matchJoinKey) == 0) {
```
[the remark this answers is not in the document]
Comment (gczsjdy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: I agree with you.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Once we advance both the streamed and buffered iterators, and call `bufferMatchingRows` at the last turn, it will advance buffered iterator until the `bufferedRow` doesn't match with current `streamedRowKey`. 

In the next turn, the call of ``advancedBufferedIterRes` here will advance buffered iterator and so the `bufferedRow` will be missed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 5]
```diff
@@ -700,38 +701,43 @@ private[joins] class SortMergeJoinScanner(
       bufferedMatches.clear()
       false
     } else if (matchJoinKey != null && keyOrdering.compare(streamedRowKey, matchJoinKey) == 0) {
-      // The new streamed row has the same join key as the previous row, so return the same matches.
+      // The new streamed row has the same join key as the previous row, so return the same
+      // matches.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Oh. I see. `advancedBufferedIterRes` is a lazy val.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Then add a comment like `Initialization at the first time reaching here`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: `This is made lazy to run the initialization only once when accessing it.`

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java [hunk 6]
```diff
@@ -182,18 +183,14 @@ public UnsafeRow next() {
         }
       };
     } catch (IOException e) {
-      cleanupResources();
       throw e;
+    } finally {
+      // Since we won't ever call next() on an empty iterator, we need to clean up resources
+      // here in order to prevent memory leaks.
+      cleanupResources();
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: This makes the resource cleaned up when we return iterator too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Add a comment like https://github.com/apache/spark/pull/19862/files#r154567168.

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java [hunk 7]
```diff
@@ -159,6 +154,12 @@ public boolean hasNext() {
         @Override
         public UnsafeRow next() {
           try {
+            if (!alreadyCalculated) {
+              while (inputIterator.hasNext()) {
+                insertRow(inputIterator.next());
+              }
+              alreadyCalculated = true;
+            }
             sortedIterator.loadNext();
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: `sortedIterator` is already assigned at L143. When you insert rows when first time to call `next`, can the `sortedIterator` correctly return sorted elements?

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java [hunk 8]
```diff
@@ -159,6 +154,12 @@ public boolean hasNext() {
         @Override
         public UnsafeRow next() {
           try {
+            if (!alreadyCalculated) {
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: When `hasNext` is called, doesn't `sortedIterator` return no element anymore since we haven't added rows into it?

Comment (gczsjdy) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: Yes. There are mistakes here, I will change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala (hunk 1 above)
Comment (gczsjdy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala]: Good advice. Thx.

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
I'm not sure we should spend time to optimize non-whole-stage-codegen path, which is only used as a safe fallback if whole stage codegen is not applicable. I think the more important thing to do is to support whole stage codegen for more queries.  cc @hvanhovell

File: sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java (hunk 7 above)
Comment (gczsjdy) [reply sql/catalyst/src/main/java/org/apache/spark/sql/execution/UnsafeExternalRowSorter.java]: Yes, you are right. Now I modified the `sortedIterator` after inserting rows. Due to I can only access an outer final field inside an inner class, so I used an array, is there better solution?

**Comment by gczsjdy** [comment]
This is actually a small change, but it can provide not small optimization for users who don't use `WholeStageCodegen`, for example there're still some users who use under 2.0 version of Spark.

Also, the behavior of codegen and non-codegen code paths are supposed to be the same, so in this way it is a 'bug'.

**Comment by gczsjdy** [comment]
cc @cloud-fan @hvanhovell  @viirya

**Comment by cloud-fan** [comment]
I don't agree this is a small change, and users using spark prior to 2.0 won't get this patch, as we don't backport performance improvement patches.

Overall this patch won't bring much benifit to Spark and may not worth the time to review it.

**Comment by gczsjdy** [comment]
@cloud-fan Ok, thanks for your time, I will close this.
